hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java

```diff
@@ -16,116 +16,135 @@
-        this.partitionResults = createHashMap(partitionCount);
+        this.targets = new HashSet<Address>(memberPartitions.keySet());
+        this.results = createHashMap(targets.size());
+        this.partitionResult = createHashMap(partitionCount);
     }
 
-    /**
-     * Executes all the operations on the partitions.
-     */
-    <T> Map<Integer, T> invoke() throws Exception {
-        ensureNotCallingFromPartitionOperationThread();
-
-        invokeOnAllPartitions();
-
-        awaitCompletion();
-
-        retryFailedPartitions();
-
-        return (Map<Integer, T>) partitionResults;
-    }
-
-    private void ensureNotCallingFromPartitionOperationThread() {
-        if (Thread.currentThread() instanceof PartitionOperationThread) {
-            throw new IllegalThreadStateException(Thread.currentThread() + " cannot make invocation on multiple partitions!");
-        }
-    }
-
-    private void invokeOnAllPartitions() {
+    <E> ICompletableFuture<E> invoke() {
         for (Map.Entry<Address, List<Integer>> mp : memberPartitions.entrySet()) {
             Address address = mp.getKey();
             List<Integer> partitions = mp.getValue();
             PartitionIteratingOperation op = new PartitionIteratingOperation(operationFactory, toIntArray(partitions));
-            Future future = operationService.createInvocationBuilder(serviceName, op, address)
-                    .setTryCount(TRY_COUNT)
-                    .setTryPauseMillis(TRY_PAUSE_MILLIS)
-                    .invoke();
-            futures.put(address, future);
+            InternalCompletableFuture<PartitionResponse> future =
+                    operationService.createInvocationBuilder(serviceName, op, address)
+                            .setTryCount(TRY_COUNT)
+                            .setTryPauseMillis(TRY_PAUSE_MILLIS)
+                            .invoke();
+            future.andThen(new SingleTargetCallback(address));
         }
+        return this;
     }
 
-    private void awaitCompletion() {
-        NodeEngineImpl nodeEngine = operationService.nodeEngine;
-        for (Map.Entry<Address, Future> response : futures.entrySet()) {
-            try {
-                Future future = response.getValue();
-                PartitionResponse result = nodeEngine.toObject(future.get());
-                result.addResults(partitionResults);
-            } catch (Throwable t) {
-                if (operationService.logger.isFinestEnabled()) {
-                    operationService.logger.finest(t);
-                } else {
-                    operationService.logger.warning(t.getMessage());
-                }
-                List<Integer> partitions = memberPartitions.get(response.getKey());
-                for (Integer partition : partitions) {
-                    partitionResults.put(partition, t);
+    public synchronized void complete(Address target, Object result) {
```

Comment:
Do we need this synchronized ? Synchronized can be problematic for a lot of reasons. One of the reasons I got rid of synchronized in code for futures is because if there is contention (and suspension) the lock gets converted to a fat lock which needs to be specially dealt with during gc.